Enable desired linters and document disabled ones

tpantelis opened the pull request:
See individual commits for details. A separate commit was created for each linter that yielded errors.

tpantelis committed:
Enable 'dupl' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'errorlint' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'exhaustive' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'gci' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

skitt commented on main.go:
```diff
@@ -290,7 +292,7 @@ func startHTTPServer() *http.Server {
 
 	go func() {
-		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+		if err := srv.ListenAndServe(); errors.Is(err, http.ErrServerClosed) {
```
This is inverting the conditional.

```suggestion
		if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
```

tpantelis replied:
Oh right - thanks for noticing that.

skitt commented on pkg/globalnet/main.go:
```diff
@@ -140,7 +141,7 @@ func startHTTPServer() *http.Server {
 
 	go func() {
-		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+		if err := srv.ListenAndServe(); errors.Is(err, http.ErrServerClosed) {
```
```suggestion
		if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
```

skitt commented on pkg/natdiscovery/natdiscovery.go:
```diff
@@ -148,7 +148,7 @@ func (nd *natDiscovery) AddEndpoint(endpoint *v1.Endpoint) {
 	// support nat discovery disabled or a remote cluster endpoint which still hasn't implemented this protocol
 	if _, err := extractNATDiscoveryPort(&endpoint.Spec); err != nil || nd.serverPort == 0 {
-		if err != errorNoNATDiscoveryPort {
+		if errors.Is(err, errorNoNATDiscoveryPort) {
```
```suggestion
		if !errors.Is(err, errorNoNATDiscoveryPort) {
```

skitt commented on pkg/routeagent_driver/handlers/kubeproxy/vxlan.go:
```diff
@@ -20,12 +20,12 @@ package kubeproxy
 	"strings"
 	"syscall"
 
+	extErrors "github.com/pkg/errors"
```
I think it would be clearer long term if we come up with a consistent set of package shorthands: _e.g._ `errors` for Golang `errors`, `exterrors` for the GH package, and `apierrors` for the K8s package.

tpantelis replied:
I use `apierrors` (or `apiErrors`) anyway. Re: the other two, we've mostly used the GH package so `errors` made sense. But with the newer APIs (eg `Is`, `As` etc) in the Golang package, we may be using that more. However I think GH will still be used much more so perhaps we should keep using `errors` for that and use `goerrors` (or `goErrors`) for the other.

Either way I think that should be a separate PR.

skitt replied:
Yeah, I was listing `apierrors` for completeness. `errors` for the GH package makes sense, with `goerrors` for the Golang package. Let’s make it a separate PR, yes.

skitt commented on pkg/routeagent_driver/handlers/ovn/bridge.go:
```diff
@@ -81,7 +81,7 @@ func (ovn *Handler) configureSubmarinerInterface() error {
 	}}
 
-	if err := ovn.netlink.AddrAdd(ovnLink, ipConfig); err != nil && err != syscall.EEXIST {
+	if err := ovn.netlink.AddrAdd(ovnLink, ipConfig); err != nil && errors.Is(err, syscall.EEXIST) {
```
```suggestion
	if err := ovn.netlink.AddrAdd(ovnLink, ipConfig); err != nil && !errors.Is(err, syscall.EEXIST) {
```

tpantelis commented:
@skitt Thanks for the first review. This is still in draft b/c I still have more linters to enable. I thought it was easier to put them all in one PR rather than handling conflicts and rebases across many PRs.

skitt commented:
> @skitt Thanks for the first review. This is still in draft b/c I still have more linters to enable. I thought it was easier to put them all in one PR rather than handling conflicts and rebases across many PRs.

You’re welcome, I agree it makes sense to enable everything in one PR! I’ll update my performance PR...

tpantelis committed:
Enable 'gocognit' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Change min-complexity for 'gocyclo' to 15 and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'godot' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'gofumpt' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'nilerr' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'nolintlint' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'predeclared' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'revive' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'testpackage' linter and fix errors

There's several test files that test package-scoped functions that were
renamed to include 'internal' so they're ignored by the linter.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'wrapcheck' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'wsl' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable 'promlinter' linter and fix errors

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Enable desired linters and document disabled ones

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

Jaanki commented on .golangci.yml:
```diff
@@ -10,81 +10,129 @@ linters-settings:
       - fieldalignment
   lll:
     line-length: 140
+  nlreturn:
```
Not needed since we are not enabling nlreturn

tpantelis replied:
I left it just in case we do but might as well remove it.
